src/Strategy/Dart/PubSpecLock.hs

```diff
@@ -0,0 +1,171 @@
+module Strategy.Dart.PubSpecLock (
+  analyzePubLockFile,
+  PackageName (..),
+  PubLockContent (..),
+  PubLockPackageMetadata (..),
+  logIgnoredPackages,
+
+  -- * for testing
+  buildGraph,
+  PubDepSource (..),
+  toDependency,
+  isSupported,
+) where
+
+import Control.Applicative (optional)
+import Control.Effect.Diagnostics (Diagnostics, context)
+import Data.Aeson.Types (
+  FromJSONKey,
+ )
+import Data.Aeson.Types qualified as AesonTypes
+import Data.Foldable (asum, for_)
+import Data.Map (Map)
+import Data.Map.Strict qualified as Map
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Yaml (FromJSON (parseJSON), (.:), (.:?))
+import Data.Yaml qualified as Yaml
+import DepTypes (
+  DepEnvironment (EnvDevelopment, EnvProduction),
+  DepType (GitType, PubType),
+  Dependency (..),
+  VerConstraint (CEq),
+ )
+import Effect.Exec (Exec, Has)
+import Effect.Logger (Logger (..), Pretty (pretty), logDebug)
+import Effect.ReadFS (ReadFS, readContentsYaml)
+import GHC.Generics (Generic)
+import Graphing (Graphing, deep, direct, empty)
+import Path
+import Types (GraphBreadth (..))
+
+newtype PackageName = PackageName {unPackageName :: Text} deriving (Show, Eq, Ord, FromJSONKey)
+newtype PubLockContent = PubLockContent {packages :: Map PackageName PubLockPackageMetadata} deriving (Show, Eq, Ord)
+
+-- | Represents Pub Dependency's metadata from lock file.
+data PubLockPackageMetadata = PubLockPackageMetadata
+  { pubLockPackageIsDirect :: Bool
+  , pubLockPackageSource :: PubDepSource
+  , pubLockPackageVersion :: Maybe VerConstraint
```

Comment:
Let's keep our domain types out of the parsed types -- especially because we always just map in `CEq`

```suggestion
  , pubLockPackageVersion :: Maybe Text
```